### DISCUSSION TITLE
communication between python and rp_varibles

### DISCUSSION THREAD

**Discussion by etiennelame** [ansys/pyfluent#3319] [Q&A] [answered]
hi everyone,
   Recently i am working with a case, where I read the outlet flow and use it to calculate outlet pressure boundary. Instead of in udf, i want to calculate it in python. 
  My idea is to create a rp-variable, calculate the outlet flow with DEFINE_ON_DEMAND and set the rp-variable value, and read the rp-variable value in python. The session class has access to the rp-variable value. 
  However, I find that the rp-variable read with python do not change, even though it is changed in udf or changed in Fluent GUI console.
  Now, to do the same thing, I have to report the flow in fluent and capture the fluent stream to capture the flow value. If I can get the rp-variable value directly, it would be more convenient.
  If you have any idea about my question, please tell me the solution. Thanks for your kindness.

**Comment by ansjschuetz** [comment, marked as answer]
Hello Etienne,
I understand that you try to modify an RP var. in a UDF and then want to query that value in Python.
For that, you need to make sure that...
-- the `RP_Set_...("...", ...);` in the UDF is done on the host process
-- you fetch the value from the host process into the cortex process using e.g. the Scheme command `(%rpgetvar '...)` (where `'...` is the name of the RP var. with a leading single quote).
I don't know whether there is a Python equivalent for the latter.
If not, you should be able to use `solver.execute_tui(r'''(%rpgetvar '...)''')`.
(Note that in this Scheme command the `%` prefix makes the difference -- it ensures that the value is fetched from the host process instead of using a cortex-side cache [which is updated by this operation].)
Hope this helps.

By the way -- from you description it sounds if you want to control the outlet static pressure boundary condition based on the outlet mass flow rate through that same outlet.
Have you considered whether any built-in boundary condition type in Fluent can do that job for you?
It would most probably be more robust numerically.
Alternatively, you may be able to use an expression to define the outlet boundary static pressure.
(I once heard something about Python expressions in Fluent, but that may just have been an enhancement request..)
Finally, if you still stick with the UDF, consider using DEFINE_PROFILE to set the outlet boundary static pressure and, in that, obtaining the outlet mass flow rate. That way, you make sure that the value can be updated every iteration, which will be essential for achieving solver convergence. (Still, using an expression may provide even higher robustness...)
And for the UDF programming, keep the parallelisation requirements in mind...

**Reply by etiennelame** [reply]
Thank you for your rely @ansjschuetz .

I will try setting RP-variable in host process，I didn't pay attention to that part. 

By the way, I am also trying to go back to UDF and using DEFINE_PROFILE. But I will recommand not updating boundary condition in every iteration in my case, where i am doing 0D-3D coupling with lumped parameter model. ODEs need to be solved in lumped parameter model, that's why i use UDF or python instead of expression, and it would be better to update bcs after enough iterations.

**Reply by cons013** [reply]
@etiennelame did you ever come up with a nice solution for this? I too am looking to try create/modify/use rp variables with a pyfluent script and a udf.c file, but am having trouble getting it to all work properly.

**Reply by etiennelame** [reply]
@cons013 This is my test case. I define a rp-variable named 'tes and set its value in UDF:
**DEFINE_ON_DEMAND(test_rp)**
**{**
**#if RP_HOST**
    **RP_Set_Real("tes2", 3.1);**
**#endif**
**}**

**DEFINE_ON_DEMAND(test_rp2)**
**{**
**#if RP_HOST
    RP_Set_Real("tes2", 2.7);
#endif
}**

Then in Python, i used the following code to return value of tes2, the returned value changes when I excute different DEMANDs:
**lispy.parse(solver.rp_vars._eval_fn("(%rpgetvar 'tes2)"))**

The library **lispy** is imported with:
**import ansys.fluent.core.filereader.lispy as lispy**

I find out this function in the  source code: **rpvars.py**

**Comment by cons013** [comment]
@etiennelame Would you also know how to create an rp-var within the pythonic/fluent side? I tried this which follows an example in the doc:

![image](https://github.com/user-attachments/assets/6d8d7bb2-efe9-4154-abe9-6110451ba056)

I.e, something similar to (rp-var-define) but without using scheme_eval.

best,
cons013

**Reply by etiennelame** [reply]
hi @cons013 , I  have not tried define rp variable in python. I would recommend **solver.scheme_eval** and see if it can send scheme command to fluent.
